services-operator/beamsqlstatementsetoperator.py

```diff
@@ -0,0 +1,282 @@
+            if not state == States.CANCELING.name or not state == States.CANCELED.name:
+                flink_util.cancel_job(logger, job_id)
+        except Exception as err:
+            raise kopf.TemporaryError(f"Error trying to cancel {namespace}/{name} with message {err}. Trying again later", 10)
+        # cancelation went through, delete job_id and update state
+        patch.status[JOB_ID] = None
+        patch.status[STATE] = States.CANCELING.name
+        raise kopf.TemporaryError(f"Waiting for confirmation of cancelation for {namespace}/{name}", 5)
+    elif state == States.CANCELING.name:
+        refresh_state(body, patch, logger)
+        if not patch.status[STATE] == States.CANCELED.name:
+            raise kopf.TemporaryError(f"Canceling, waiting for final confirmation of cancelation for {namespace}/{name}", 5)
+    kopf.info(body, reason="deleting", message=f" {namespace}/{name} cancelled and ready for deletion")
+    logger.info(f" {namespace}/{name} cancelled and ready for deletion")
+
+@kopf.index('oisp.org', "v1alpha1", "beamsqltables")
+def beamsqltables(name: str, namespace: str, body: kopf.Body, **_):
+    return {(namespace, name): body}
+
+
+@kopf.timer("oisp.org", "v1alpha1", "beamsqlstatementsets", interval=timer_interval_seconds, backoff=timer_backoff_seconds)
+def updates(beamsqltables: kopf.Index, stopped, patch, logger, body, spec, status, **kwargs):
+    """
+    Managaging the main lifecycle of the beamsqlstatementset crd
+    Current state is stored under
+    status:
+        state: STATE
+        job_id: string
+    STATE can be
+        - INITIALIZED - resource is ready to deploy job. job_id: None
+        - DEPLOYING - resource has been deployed, job_id: flink_id
+        - RUNNING - resource is running job_id: flink_id
+        - FAILED - resource is in failed state
+        - DEPLOYMENT_FAILURE - something went wrong while operator tried to deploy (e.g. server returned 500)
+        - CANCELED - resource has been canceled
+        - CANCELING - resource is in cancelling process
+        - UNKNOWN - resource cannot be monitored
+
+    Transitions:
+        - undefined/INITIALIZED => DEPLOYING
+        - INITIALIZED => DEPLOYMENT_FAILURE
+        - DEPLOYMENT_FAILURE => DEPLOYING
+        - DEPLOYING/UNKNOWN => FLINK_STATES(RUNNING/FAILED/CANCELLED/CANCELLING)
+        - delete resource => CANCELLING
+        - delete done => CANCELLED
+    Currently, cancelled state is not recovered
+    """
+    namespace = body['metadata'].get("namespace")
+    name = body['metadata'].get("name")
+    state = body['status'].get(STATE)
+    logger.debug(f"Triggered updates for {namespace}/{name} with state {state}")
+
+    if state == States.INITIALIZED.name or state == States.DEPLOYMENT_FAILURE.name:
+        # deploying
+        logger.debug(f"Deyploying {namespace}/{name}")
+
+        # get first all table ddls
+        # get inputTable and outputTable
+        ddls = f"SET pipeline.name = '{namespace}/{name}';\n"
```

Comment:
Can there be an SQL Injection here? (Depending on who we allow to create queries long term, it might make sense to sanity check namespace and name)